### PR TITLE
Track stack_numeric_size per register to allow verifying programs that join multiple stack regions

### PR DESCRIPTION
**Pull request by dthaler**
This change allows the verifier to deal with the twostackvars sample in
https://github.com/vbpf/ebpf-samples/pull/21
where "stack_numeric_size" is tracked per variable that contains a stack pointer.
The case in twostackvars.o happens commonly in practice due to register spilling, such as in
the Cilium layer-4 load balancer eBPF programs (which are used in the work
mentioned at https://cloudblogs.microsoft.com/opensource/2022/02/22/getting-linux-based-ebpf-programs-to-run-with-ebpf-for-windows/ which refers to PREVAIL contributions, which are this PR and the now-merged PR #289).

One way to think of the stack_numeric_size is similar to a typedef of an integer pointer,
e.g., is it a uint8_t* or a uint16_t* or a uint64_t* etc.
Since it is common in code to form a larger integer value from
component parts (e.g., in htons, ntohl, etc.) it is important to
be able to expand/contract the size based on observations, while
still tracking whether one can safely dereference the variable
to get a number of a given length.  So stack_numeric_size holds the number
of continuous bytes that are known to be numeric.  For example,
a stack_numeric_size of 0 means that it is not even a uint8_t*.
As such, this is more efficient and than say trying to create T_STACK_INT8_PTR, T_STACK_INT16_PTR, etc. where tracking changing of types based on specific bytes being modified would be far more complex.

A different way of thinking about stack_numeric_size is that it's similar to shared_region_size (which is specific to shared type pointers rather than stack type pointers) except that stack_numeric_size is the size of the numeric-only region instead of the full region.

The stack_numeric_size for all stack type variables is always updated whenever the stack is touched, since the numeric region size might have changed.

### CODE REVIEW

**Comment by elazarg** [comment]
I'll do my best to get to it this Thursday.

File: src/crab/ebpf_domain.cpp
Comment (elazarg) [inline src/crab/ebpf_domain.cpp:1556]: Seems incorrect. 
```suggestion
                        auto ub = m_inv.eval_interval(src.value).lb().number();
                        auto lb = m_inv.eval_interval(src.value).lb().number();
                        if (ub && ub.value().fits_sint() && ((int)ub.value() < 0) && lb && lb.value().fits_sint())
                            // If it must be a negative number, narrow the numeric size.
                            apply(inv, crab::arith_binop_t::ADD, dst.stack_numeric_size, dst.stack_numeric_size, (int)lb, false);
```

**Review by elazarg — CHANGES_REQUESTED** [review]
Encoding "minimum known size" into Zones is less straightforward than it seems. For example, the join of `{min_size=5}` and `{min_size=4}` is actually `min_size=T` instead of the expected `{min_size=4}`. It can be fixed, but it would be either a hack, or a separate numerical domain.

In the example above, it should be `{size=[5,oo]}` U `{size=[4,oo]}` = `{size=[4,oo]}` (since join is `min` for lower bounds). Which means we should track `next_non_numeric` or something like that.

---
Most of the inline comments are minor, except the subtraction.

As an idea, it might be useful to track numeric size for all pointer types.

**Comment by caballa** [comment]
@elazarg : the join of {min_size=5} and {min_size=4} should return {4 <= min_size <=5 } from which you can extract the lower bound 4.

**Comment by elazarg** [comment]
> @elazarg : the join of {min_size=5} and {min_size=4} should return {4 <= min_size <=5 } from which you can extract the lower bound 4.

You are correct of course. I've confused myself.

**Comment by dthaler** [comment]
> As an idea, it might be useful to track numeric size for all pointer types.

Other pointer types (except T_CTX) are all numeric-only so the numeric size is the region size which is already tracked.

**Comment by elazarg** [comment]
> > As an idea, it might be useful to track numeric size for all pointer types.
> 
> Other pointer types (except T_CTX) are all numeric-only so the numeric size is the region size which is already tracked.

That's why an exact encoding would be infinity, not top. I think it can add precision, or make it more easily fall out of the implementation.

File: src/crab/ebpf_domain.cpp
Comment (dthaler) [reply src/crab/ebpf_domain.cpp:359]: I can't remember since this PR was started long ago.  It does the TODO in the old code.  Let me know if you want it reverted, but since the TODO was there I think it's useful.

Comment (elazarg) [reply src/crab/ebpf_domain.cpp:359]: No need to revert

File: src/crab/ebpf_domain.cpp
Comment (dthaler) [reply src/crab/ebpf_domain.cpp]: Done

File: src/crab/ebpf_domain.cpp
Comment (dthaler) [reply src/crab/ebpf_domain.cpp]: Done

File: src/crab/array_domain.cpp
Comment (dthaler) [reply src/crab/array_domain.cpp]: Agreed, done.

File: src/crab/ebpf_domain.cpp
Comment (dthaler) [reply src/crab/ebpf_domain.cpp:1228]: Agree in separate PR

File: src/crab/var_factory.cpp
Comment (dthaler) [reply src/crab/var_factory.cpp:35]: I'm interpreting this is just a comment about possible future optimization, not any request for change in this PR.

File: src/crab/ebpf_domain.cpp
Comment (dthaler) [reply src/crab/ebpf_domain.cpp]: Replaced with a comment

File: src/crab/ebpf_domain.cpp
Comment (dthaler) [reply src/crab/ebpf_domain.cpp]: Updated (this location was simplified further since the result is T_NUM) but I did your suggestion in another location in this file where it did matter.

File: src/crab/ebpf_domain.cpp
Comment (elazarg) [inline src/crab/ebpf_domain.cpp:1518]:
```suggestion
                                    auto ub = m_inv.eval_interval(src.value).ub().number();
```

File: src/crab/ebpf_domain.cpp
Comment (elazarg) [inline src/crab/ebpf_domain.cpp:1581]:
```suggestion
                            auto ub = m_inv.eval_interval(src.value).ub().number();
```

File: src/crab/ebpf_domain.cpp
Comment (dthaler) [reply src/crab/ebpf_domain.cpp]: Fixed

File: src/crab/ebpf_domain.cpp
Comment (dthaler) [reply src/crab/ebpf_domain.cpp]: Fixed

File: src/crab/ebpf_domain.cpp
Comment (elazarg) [reply src/crab/ebpf_domain.cpp:1521]: Using `dsl_syntax`, this can be written more directly as
```
using crab::dsl_syntax;
if (m_inv.intersect(src.value < 0)) havoc();
else ...
```

**Review by elazarg — CHANGES_REQUESTED** [review]
- [ ] I am not sure the handling of `r += r` is correct, see the comments there. This is the "request changes" part.

---

General comment: this feature is more delicate than other features, for two related reasons: aliasing and reduction.

Aliasing means the information about one variable is invalidated by an operation on another variable. This is required for soundness. Luckily, this only happens on `do_store_stack()`, though in the future may also happen on CTX. I see no issue there.

Reduction means we can propagate information about one variable (say `r1.stack_numeric_size`) to another "variable" (num_bytes) and vice versa. This is needed for precision. I think it can be useful in many places, but IIUC this PR does it only on `recompute_stack_numeric_size()`, which is called after a pointer is moved backwards. Since it is only about precision, this should not block the PR.

My guess is that this feature may not play well with the future algebraic domain I want to use for the types. But I have no concrete examples for that.

---

Most of the comments are one large suggestion: I think the PR can be written using linear expressions and constraints, which can make it both more readable and more general. I've tried some (but not all) of those suggestions, and they seem to work.

In the code, I think `inv.entail(cst)` can be read as `inv.necessarily(cst)` and `inv.intersect(cst)` can be read as `inv.maybe(cst)`. Admittedly each is implemented by copying the state, which is slow, but optimizations may be implemented inside `SplitDBM`, and are implemented for `entail()`.

One culprit of using `entail`/`intersect` and `+=` (or `assume()`) is that `stack_numeric_size=8` becomes `stack_numeric_size=[8, +oo]`, and TOP is `[0, +oo]` which creeps everywhere, both clattering the output and incurring some needless performance cost. The former may be addressed in pretty-printing phase, and the latter can be improved by somehow changing it to TOP (I'm not sure it is as trivial as it sounds).

---

@dthaler if you prefer keeping the current approach (extracting lower- and upper-bounds), then once my issues with ADD are settled I can merge, and try to implement this approach myself later.

@caballa I'd appreciate a quick look at my suggestions, particularly the use of `entail` and `intersect` - just to verify I'm not confusing myself again :) Of course, a deeper review would be very helpful.

**Comment by dthaler** [comment]
> @dthaler if you prefer keeping the current approach (extracting lower- and upper-bounds), then once my issues with ADD are settled I can merge, and try to implement this approach myself later.

Yes that's what I'd prefer so we can switch back to using the upstream repo instead of my fork, sooner than later.

**Comment by caballa** [comment]
@elazarg: we should have a private conversation about what you intend to achieve because I'm not sure I follow but let me try.

My understanding is that you are worried that the some operations delegate to interval reasoning but that can produce  imprecise results. Just to exemplify what I mean, let's say that we want to know whether x and y are definitely equal. The precise way to ask the numerical abstract domain is: 

```
inv+= (linear_constraint_t(x) != linear_constraint(y))
return inv.is_bottom();
```

which is what `entails` does but it copies `inv` as you mentioned to ensure that the effects of adding the constraint are undone after the call to `is_bottom`.

Alternatively, we could have an imprecise version of that:

```
intv_x = inv.to_interval(x);
intv_y = inv.to_interval(y);

if (intv_x. lb().is_finite() && intv_x.ub().is_finite() && intv_y.lb().is_finite() && intv_y.ub().is_finite()) {
     return intv_x.lb().number() == intv_y.lb().number() && intv_x.ub().number() == intv_y.ub().number()
}  else {
    return false;
}
```

Are you suggesting to have new abstract operations that can hide whether interval reasoning is used or not to avoid unexpectedly precision losses? If yes, I think it's a good idea but I don't really follow what is the semantics of your new `entails` and `intersect` operations.

File: src/crab/ebpf_domain.cpp
Comment (dthaler) [reply src/crab/ebpf_domain.cpp:1384]: Tried your suggestion and it results in a bunch of other invariants being added unrelated to registers in the yaml test case.  E.g.:
```
Filters: [yaml]
test case: stack assign immediate
unseen inv: []
unexpected inv: [
    r1.stack_numeric_size=[0, +oo],
    r10.stack_numeric_size=[0, +oo],
    r2.stack_numeric_size=[0, +oo]]
```

I haven't tried to track this down, as I hope it isn't a blocker, so maybe it can be done in a future PR.

File: test-data/stack.yaml
Comment (dthaler) [reply test-data/stack.yaml]: Fixed name

File: src/crab/ebpf_domain.cpp
Comment (dthaler) [reply src/crab/ebpf_domain.cpp]: Updated as suggested

File: src/crab/ebpf_domain.cpp
Comment (dthaler) [reply src/crab/ebpf_domain.cpp]: Updated as suggested

File: src/crab/ebpf_domain.cpp
Comment (dthaler) [reply src/crab/ebpf_domain.cpp]: Updated as suggested, but did not translate to the ADD case here.  Let me know if you think it's required, but it doesn't seem like a blocker to me.

File: src/crab/ebpf_domain.cpp
Comment (dthaler) [reply src/crab/ebpf_domain.cpp]: Done

File: src/crab/ebpf_domain.cpp
Comment (dthaler) [reply src/crab/ebpf_domain.cpp]: Done

File: src/crab/ebpf_domain.cpp
Comment (dthaler) [reply src/crab/ebpf_domain.cpp]: Updated as suggested

**Comment by elazarg** [comment]
@caballa let's talk have a chat. I think it's outside the scope here anyway since none of this is going to be implemented in this PR.

File: src/crab/ebpf_domain.cpp
Comment (elazarg) [reply src/crab/ebpf_domain.cpp:1381]: Not a blocker. These are the invariants I've mentioned (they are just TOP, really).

File: src/crab/ebpf_domain.cpp
Comment (elazarg) [inline src/crab/ebpf_domain.cpp:1571]: Shouldn't this be `T_STACK`?

File: src/crab/ebpf_domain.cpp
Comment (elazarg) [inline src/crab/ebpf_domain.cpp:1404]: (Just a TODO for future PR):

The warning given by codecov led me to notice that here is the only case where stack_numeric_size is not TOP when `recompute_stack_numeric_size()` is called, and I'm not sure it is particularly needed. If recompute_stack_numeric_size() is moved inside `if (imm < 0)` then it can do the havoc() internally - or anything more sophisticated that might be possible.

**Comment by elazarg** [comment]
LGTM except the `T_PACKET` typo. I'd fix that myself but I think the DCO check may fail. I'll merge once this is fixed.

File: src/crab/ebpf_domain.cpp
Comment (dthaler) [reply src/crab/ebpf_domain.cpp]: Yes! Fixed

File: src/crab/ebpf_domain.cpp
Comment (dthaler) [reply src/crab/ebpf_domain.cpp:1404]: For code coverage, I added a test for this case, but the test just reproduces the issue in #321

File: src/crab/ebpf_domain.cpp
Comment (dthaler) [inline src/crab/ebpf_domain.cpp:1570]: For code coverage, I added a test for this case, but the test just reproduces the issue in https://github.com/vbpf/ebpf-verifier/issues/321

**Review by elazarg — APPROVED** [review]

**Pull request merged**
